rstudio/rstudio
# Feature/session suspend info

### Intent

The Open Source PR for https://github.com/rstudio/rstudio-pro/issues/2618. This code tracks IDE operations that block a session from timing out and suspending, and reports these "blocking ops" to the user when a session attempts to suspend, but cannot.

### Approach

The code that manages suspension is scattered through several locations. Wherever was possible, the suspend-related logic was moved into `SessionSuspend.hpp/cpp`. Internally `SessionSuspend` will track any operations that can block timeout suspension to later be reported if and when the session attempts to suspend because of a timeout. It is the responsibility of external code to report its own suspend-blocking operations to `SessionSuspend` via `SessionSuspend::[add/remove/check]BlockingOps()` calls.

~~To minimize noise to the user an indicator that the session attempted to suspend, but was blocked, is only displayed once a suspend attempt was made.~~ When a user starts an activity that will block automatic suspension, an icon will appear in the R console toolbar within 5 seconds to indicate that the session will remain active until the task completes, if ever. The indicator will remain visible until all blocking operations have stopped, updating dynamically as blocking ops start or stop. As long as the indicator is visible, the session will not attempt to suspend regardless the amount of time that passes. Once all blocking operations have stopped the indicator will disappear, the timeout will restart to whatever value is set by the `session-timeout-minutes` option, and the process repeats.

To minimize noise to the logs ~~a warning~~ an informational-level log that a session attempted to suspend is only emitted once ~~per session suspend attempt~~ when both conditions are true:

1. Auto suspension is actively being blocked by some user activity
2. The above activity ran longer than whatever value is set by `session-timeout-minutes`

Aka, if the session would have normally tried to auto suspend, but couldn't because auto suspension was blocked. The warning lists the blocking operations as well as the session id, username, and the value of `session-timeout-minutes`. 

None of this additional logic will interfere with session suspend request RPCs or forced session suspends.

A "Session Suspended" Icon was added to the R console toolbar. It will display whenever the session is suspended, similar to how the memory usage widget grays out when a session suspends.

### Automated Tests

This PR doesn't include automated testing. A basic selenium testing suit might include:

1. Start a new R session
2. Create one of the conditions that will prevent R sessions from suspending
3. Wait for session timeout
4. Session does not suspend
5. Mouse over/title of suspend blocked icon should match the condition we created in step 2
6. Stop/remove the blocking op
7. suspend blocked icon should disappear

We will likely want to test each of the suspend-blocking operations separately, as well as some(or all) together to make sure multiple reasons a session won’t suspend are also displayed.


### QA Notes

See the testing section for how to reproduce. Other things to consider:

- Requesting a session to suspend from the GUI (Session->Suspend R Session) should still force the session to suspend, even if the suspend blocked icon is displaying
- This should not change any functionality on the desktop app, as the desktop app does not suspend

### Checklist
